### ISSUE TITLE
Usage of symbols Cmd.getState and Cmd.dispatch is hard to type-check

### ISSUE THREAD

**Issue by laszlopandy** [redux-loop/redux-loop#233]
The symbols `Cmd.getState` and `Cmd.dispatch` are passed to `Cmd.run` in order to indicate that the special arguments `getState` and `dispatch` should be passed to the given function. I believe we should brainstorm more ideas and try to improve this API.

## Current issues
 * Passing symbols in an array to indicate special arguments (AFAIK) is not idiomatic JavaScript. I have never seen this pattern used in other JavaScript libraries. It may be difficult for some first-time users to grasp
 * It is difficult (or impossible) to type check the argument types in TypeScript (see https://github.com/redux-loop/redux-loop/pull/222). This is not a deal-breaker on it's own. Redux-loop is not a TypeScript library, and TypeScript cannot express all JavaScript APIs. But if we can solve this as well it would be great for our TypeScript users.
 * I don't use Flow but my understanding of the Flow type-definitions is the issue is the same as in TypeScript.

## Current ideas
Feel free to add more ideas below as they arise.
#### 1. Separate `Cmd.run` into two:
`Cmd.run` would works just like it does currently. A new method, `Cmd.runWithStore` would *always* pass `getState` and `dispatch` as the first two arguments.
```js
Cmd.runWithStore(
  (dispatch, getState, myArg) => {
    /* side effects here */
  },
  {
    args: [myArg],
  });
```

#### 2. Return a function to access store functions
If you want to use `Cmd.run` without `getState` or `dispatch` simply return a Promise or a value. If you want to use those special argument, return a function. This is very similar to how redux-thunk allows returning a function in an action creator.
```js
Cmd.run(
  (myArg) => (dispatch, getState) => {
    /* side effects here */
  },
  {
    args: [myArg],
  });
```

**Comment by lorefnon** [comment]
We have the following minimal helper function that we use in most of our reducers to work around the type-safety issue of `Cmd.run`: 

```ts
export type CmdFn<TArgs extends any[] = []> = (
    getState: () => RootState, 
    dispatch: Dispatch, 
    ...args: TArgs
) => void

export const runCmd =
    <TArgs extends any[]>(
        fn: CmdFn<TArgs>,
        ...args: TArgs
    ) => Cmd.run(fn, {
        args: [Cmd.getState, Cmd.dispatch, ...args]
    });
```

All Cmd handler functions just receive `getState` and `dispatch` and ignore if not needed. success/failure actions are just directly dispatched from within the function (this is anyways rare in our app because its an offline-first application).

**Comment by bdwain** [comment]
Yea I recognize that the API is not so typical, but I think it's an example of how javascript can enable a lot of flexibility by being so dynamic. It does make type checking harder, but I think the flexibility is worth it. I don't really like think redux-loop should be dictating the ordering of a side effect's parameters. The Cmd API was designed around making side effects generic and not tied to redux-loop. 

Also, doesn't #222 solve the type safety issue anyway?

**Comment by lorefnon** [comment]
> Also, doesn't #222 solve the type safety issue anyway?

Yes. I'd be really happy if that PR is merged.

I just wanted to share a simple workaround that worked for me, that people who care about type safety can use in the meanwhile.

**Comment by bdwain** [comment]
awesome yea I think that's a good workaround. I need help reviewing #222 since I'm not too familiar with typescript, but I'd love to get it merged.

**Comment by laszlopandy** [comment]
We should definitely merge #222 in the meantime; it is a non-breaking change and will help all TypeScript users. I will try to work on it soon. 

@bdwain I have a clarifying question: I agree that the current API is the most flexible, but I don't understand why that flexibility is important. Specifically I don't understand why you say "I don't really like think redux-loop should be dictating the ordering of a side effect's parameters."
Surely any function that is uses `getState()` or `dispatch()` would have been written for redux-loop no?

Perhaps you meant that the function might have been written with redux in mind but not redux-loop specifically. But in that case shouldn't we align the argument convention with the redux-thunk way of returning a function (option #2 in issue description)? What am I missing here?

**Comment by bdwain** [comment]
Doesn't #222 require a higher version of typescript? if so, it would have to be breaking. Though since #235 is breaking too, it doesn't really matter that much.

And yes, the idea was to tie it to redux, not redux-loop, since we have to tie it to redux or you can't dispatch an action or get your state. This was kind of a design goal of the Cmd API, as the old Effects API required you to return the actions that you wanted dispatched, so everything was designed specifically for redux-loop.

As far as option 2 goes, I don't think it would be so simple. We'd either have to always wrap the function in a second function (which I'd be strongly against) or you would have to enable it with an option to Cmd.run, which isn't terrible, but is still big refactor needed for any app using redux-loop.

Considering #222 introduced type safety with the current API anyway, what's the motivation for this change? I'd rather not force a large change to the API that would force a significant refactor on  all users of the library  if we can avoid it, especially if the only benefit is to simplify the types we maintain.

**Comment by laszlopandy** [comment]
@bdwain 
I'm not sure why I care so much about this issue. Perhaps it's just that I find the current Cmd.run cumbersome to use and when I was new to it I was really confused. 

I disagree about option 2. Assuming that currently there is no reason for a function used in Cmd.run to return another function, then we can easily support both modes simultaneously. 
```js
// inside handleRunCmd
    ...
    let result = cmd.func(...getMappedCmdArgs(cmd.args, dispatch, getState));
    if (typeof result === 'function') {
        result = result(dispatch, getState);
    }
    ...
```
Option 2 would be fully backwards compatible, and could even co-exist with the `Cmd.dispatch` API (although in general it's probably a bad idea for there to be two ways to do the same thing).

And if someone really needed to dispatch an action with a function in it, they could still return `Promise.resolve(myFunction)`.

**Comment by bdwain** [comment]
You may not control the definition of the function you are calling though. It may be controlled by a library you are using. An example of a pretty common pattern  that this would break would be a subscribe function that returns an unsubscribe function. If we saw that you returned a function and called it, it would immediately unsubscribe you. You might not even be planning on dispatching any actions  with the  function in it if you have no plans to cancel the subscription.

```js
import {subscribeToFoo} from 'third-party-library';

return Cmd.run(subscribeToFoo, {args: [123]})
```

If subscribeToFoo returns an unsubscribe function, you're stuck, as you will immediately unsubscribe.

Why don't we see how #222 works out, and if there are still issues with the type definitions or people are finding the API confusing, we can evaluate our options. I haven't heard many complaints about the current design besides the type safety, so I'd rather keep it to avoid churn if we can.

**Comment by laszlopandy** [comment]
Yes I see what you mean. I'm happy to revisit this in the future -- it was intended to be a open-ended discussion and not a search for an immediate solution. 

Could we leave this issue open in the meantime, to serve as documentation of the trade-offs here? At least until we have a design document on the choices we've made on this API.

**Comment by bdwain** [comment]
sure

**Comment by on3iro** [comment]
Hey guys, unfortunately i wasn't able to follow the whole issue, but I just updated redux loop and I think I have a typing issue which is related and I am not quite sure how to best resolve it:

lets say a handler of my reducer looks like this:

```ts
export const createExpedition = (
  state: State,
  action: ReturnType<typeof actions.createExpedition>
) => {
  return loop(
    state,
    Cmd.run(sideEffects.createExpedition, {
      args: [Cmd.getState, action.payload.baseConfig],
      successActionCreator: actions.createExpeditionSuccess,
    })
  )
}
```

Now I would like to type the `createExpedition()` sideEffect. Before I updated redux-loop the following did work:

```ts
export const createExpedition = (
  getState: () => RootState,
  baseConfig: BaseConfig
): types.Expedition => {
  const state = getState()
  const { expeditionConfig } = baseConfig
  ...
}
```

But now I get an error that `Cmd.getState` is not compatible to `() => RootState` because its a symbol. What would be the best way to get around this?

Thanks in advance

**Comment by bdwain** [comment]
@on3iro I use these types for the values from Cmd.getState and Cmd.dispatch

```ts
// my-app/redux-helpers.ts
import { Cmd } from 'redux-loop';

export type GetStateFn = <R = RootState>() => R;
export type DispatchFn = Cmd.Dispatch;


export const createExpedition = (
  getState: GetStateFn,
  baseConfig: BaseConfig
): types.Expedition => {
  const state = getState()
  const { expeditionConfig } = baseConfig
  ...
}

```

**Comment by on3iro** [comment]
@bdwain awesome thanks for the quick answer - i'll try this asap (probably later today) and will let you know if this works :)

**Comment by on3iro** [comment]
@bdwain the `GetStateFn` type seems to work like a charm.
